src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,158 @@
-
+        checkIndex(index, size);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> nodeUnderIndex = findNode(index);
+            Node<T> newNode = new Node<>(nodeUnderIndex.prev, value, nodeUnderIndex);
+            nodeUnderIndex.prev = newNode;
+            if (newNode.prev == null) {
+                head = newNode;
+            } else {
+                newNode.prev.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size - 1);
+        Node<T> removeNode = findNode(index);
+        T removeValue = removeNode.item;
+        removeNode.item = value;
+        return removeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        Node<T> removeNode = findNode(index);
+        size--;
+        return unlink(removeNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (object == currentNode.item || (object != null && object.equals(currentNode.item))) {
+                remove(i);
+                return true;
+            } else {
+                currentNode = currentNode.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public T unlink(Node<T> removeNode) {
+        Node<T> nextNode = removeNode.next;
+        Node<T> prevNode = removeNode.prev;
+        if (removeNode == head && removeNode == tail) {
+            head = null;
+            tail = null;
+        } else if (removeNode == tail) {
+            prevNode.next = null;
+            tail = prevNode;
+        } else if (removeNode == head) {
+            nextNode.prev = prevNode;
+            head = nextNode;
+        } else {
+            nextNode.prev = prevNode;
+            prevNode.next = nextNode;
+        }
+        removeNode.next = null;
+        removeNode.prev = null;
+        T removeItem = removeNode.item;
+        removeNode.item = null;
+        return removeItem;
+    }
+
+    private void checkIndex(int index, int lastIndex) {
+        if (index < 0 || index > lastIndex) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private Node<T> findNode(int index) {
+        if (index <= size / 2) {
+            Node<T> currentNode = head;
```

Comment:
you can initialise currentNode before if, because this node is used also in else
```suggestion
    private Node<T> findNode(int index) {
        Node<T> currentNode;
        if (index <= size / 2) {
           currentNode = head;
```

Reply:
done